q2_longitudinal/_utilities.py

```diff
@@ -400,40 +391,84 @@ def _regplot_subplots_from_dataframe(state_column, metric, metadata,
+    states = sorted(metadata[state_column].unique())
+    fig_count = len(groups) + 1
+    fig_height = fig_count * 6
+    chart, axes = plt.subplots(fig_count, figsize=(6, fig_height))
 
     # determine x tick interval: autoscale so that ≤ 30 labels appear
     xtick_interval = _set_xtick_interval(xtick_interval, states)
 
-    # plot all groups together, compare variances
-    c, global_mean, global_std = _control_chart(
-        state_column, metric, metadata, group_column, ci=ci, palette=palette,
-        plot_control_limits=plot_control_limits, ax=axes[0],
-        xtick_interval=xtick_interval)
-    c.set_title('Group volatility comparison plot')
-    c = _set_xticks(c, metadata, state_column, states, xtick_interval)
-
     # plot individual groups' control charts
     colors = cycle(sns.color_palette(palette, n_colors=len(groups)))
-    for num in range(len(groups)):
-        group = groups[num]
-        group_md = metadata[metadata[group_column] == group]
+    cmap = {}
+    num = 1
+    for group, group_md in metadata.groupby(group_column):
+        color = next(colors)
+        cmap[group] = color
         c, gm, gs = _control_chart(
             state_column, metric, group_md, None, ci=ci, legend=False,
-            color=next(colors), plot_control_limits=True, ax=axes[num + 1],
+            color=color, plot_control_limits=plot_control_limits, ax=axes[num],
             palette=None, xtick_interval=xtick_interval)
         c.set_title('{0}: {1}'.format(group_column, group))
+        if spaghetti != 'no':
+            # plot group's sphaghetti on main plot and current subplot
+            for ax in [0, num]:
+                c = _make_spaghetti(
+                    group_md, state_column, metric, individual_id_column,
+                    states, ax=axes[ax], color=color, alpha=0.3,
+                    spaghetti=spaghetti)
         c = _set_xticks(c, group_md, state_column, states, xtick_interval)
+        axes[num].set_yscale(yscale)
+        num += 1
+
+    # plot all groups together, compare variances
+    c, global_mean, global_std = _control_chart(
+        state_column, metric, metadata, group_column, ci=ci, palette=cmap,
+        plot_control_limits=plot_control_limits, ax=axes[0],
+        xtick_interval=xtick_interval)
+    c.set_title('Group volatility comparison plot')
+    c = _set_xticks(c, metadata, state_column, states, xtick_interval)
+
+    c.set_yscale(yscale)
+
     plt.tight_layout()
 
     return chart, global_mean, global_std
 
 
+def _make_spaghetti(metadata, state_column, metric, individual_id_column,
+                    states, ax, color=None, alpha=1.0, spaghetti='no'):
+    for ind, ind_data in metadata.groupby(individual_id_column):
+        # optionally plot mean of replicates
+        if spaghetti == 'mean':
+            ind_data = ind_data.groupby(state_column).mean()
+            # add state_column back into dataframe since we use it for plotting
+            ind_data[state_column] = ind_data.index
```

Comment:
Is there a way to restucture this to avoid mutation? It seems like you could probably pass the data into `_adjust_xticks`, but because this mutates, I'm not sure if there is code relying on this side-effect elsewhere.

Reply:
I think I'm not following... the mutation is intentional here, to average the data before passing to `_get_xticks` if "mean" is desired. I could do this differently, but with what seems to me like unnecessary extra code (e.g., add an `else` statement)

The mutation can't cause issues elsewhere, since `ind_data` is generated and consumed only within this `for` loop.

But perhaps I misunderstand.

Reply:
I had missed the lifetime of `ind_data` on the first pass, you're right that this probably isn't so bad.

But I think you could simplify things a bit if you passed `ind_data.index` (this is a pd.Index object right?) directly to `_get_xticks` (and `ind_data[state_column]` otherwise) which would also make `_get_xticks` simpler (or even kind of unnecessary) as it doesn't need the entire dataframe for anything, just the series (and `states`).

Here's a potential way to reduce it (I think) and makes `_get_xticks` unecessary:
```python
altered_states = ind_data[state_column]
if spaghetti == 'mean':
    altered_states = ind_data.groupby(state_column).mean().index
if states is not None:
    altered_states = altered_states.apply(states.index)
```
I am pretty sure the above is equivalent to what is here.

Reply:
nice... I needed to rearrange and tweak this a bit (your suggestion discards the mean metric), but this definitely simplifies things.

Reply:
I see what happens now, I missed the reassignment of `ind_data` being super important, which makes sense in retrospect. The reader just needs to understand the API of `df.groupby`'s result  (which I hadn't yet).